xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -0,0 +1,342 @@
+      this.priority = priority;
+    }
+
+    static LocalityLbEndpoints fromEnvoyProtoLocalityLbEndpoints(
+        io.envoyproxy.envoy.api.v2.endpoint.LocalityLbEndpoints proto) {
+      List<LbEndpoint> endpoints = new ArrayList<>(proto.getLbEndpointsCount());
+      for (io.envoyproxy.envoy.api.v2.endpoint.LbEndpoint endpoint : proto.getLbEndpointsList()) {
+        endpoints.add(LbEndpoint.fromEnvoyProtoLbEndpoint(endpoint));
+      }
+      return
+          new LocalityLbEndpoints(
+              endpoints,
+              proto.getLoadBalancingWeight().getValue(),
+              proto.getPriority());
+    }
+
+    List<LbEndpoint> getEndpoints() {
+      return Collections.unmodifiableList(endpoints);
+    }
+
+    int getLocalityWeight() {
+      return localityWeight;
+    }
+
+    int getPriority() {
+      return priority;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      LocalityLbEndpoints that = (LocalityLbEndpoints) o;
+      return localityWeight == that.localityWeight
+          && priority == that.priority
+          && Objects.equal(endpoints, that.endpoints);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(endpoints, localityWeight, priority);
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("endpoints", endpoints)
+          .add("localityWeight", localityWeight)
+          .add("priority", priority)
+          .toString();
+    }
+  }
+
+  /**
+   * See corresponding Envoy proto message {@link io.envoyproxy.envoy.api.v2.endpoint.LbEndpoint}.
+   */
+  static final class LbEndpoint {
+    private final EquivalentAddressGroup eag;
+    private final int loadBalancingWeight;
+    private final boolean isHealthy;
+
+    /** Must only be used for testing. */
+    // TODO(chengyuanzhang): migrate tests to use LbEndpoint(EquivalentAddressGroup, int, boolean)
+    //  and should add tests for LB policies dealing with unhealthy endpoints.
+    @VisibleForTesting
+    LbEndpoint(EquivalentAddressGroup eag, int loadBalancingWeight) {
+      this(eag, loadBalancingWeight, true);
+    }
+
+    private LbEndpoint(EquivalentAddressGroup eag, int loadBalancingWeight, boolean isHealthy) {
+      this.eag = eag;
+      this.loadBalancingWeight = loadBalancingWeight;
+      this.isHealthy = isHealthy;
+    }
+
+    static LbEndpoint fromEnvoyProtoLbEndpoint(
+        io.envoyproxy.envoy.api.v2.endpoint.LbEndpoint proto) {
+      io.envoyproxy.envoy.api.v2.core.SocketAddress socketAddress =
+          proto.getEndpoint().getAddress().getSocketAddress();
+      return
+          new LbEndpoint(
+              new EquivalentAddressGroup(
+                  ImmutableList.<java.net.SocketAddress>of(
+                      new InetSocketAddress(socketAddress.getAddress(),
```

Comment:
nit: too many levels of indentation. Declare a local variable `inetSocketAddress` may help.

Reply:
Done.